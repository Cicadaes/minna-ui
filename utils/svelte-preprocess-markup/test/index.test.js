/* eslint-env browser */

'use strict';

const fs = require('fs');
const { promisify } = require('util');
const { preprocess, create } = require('svelte');
const preprocessMarkup = require('../index.js');

const readFile = promisify(fs.readFile);

// don't require() components to avoid Jest transform
const componentPath = require.resolve('@minna-ui/jest-config/fixtures/TestComponent.html');

const preprocessOpts = {
  markup: preprocessMarkup(),
  style: () => ({ code: '/*discarded*/' }),
};
const preprocessOptsUnsafe = {
  markup: preprocessMarkup({ unsafe: true }),
  style: () => ({ code: '/*discarded*/' }),
};
const svelteOpts = {
  filename: 'TestComponent.html',
  name: 'TestComponent',
  onwarn(warning, onwarn) {
    /* istanbul ignore if */
    if (process.env.NODE_ENV !== 'development' && !/A11y:/.test(warning.message)) {
      onwarn(warning);
    }
  },
};

const sourceSimple = `
<div id="target">
  test

  </div>

    <hr class="one two three">

  <input type="text" name="test" class ="" readonly="readonly">

      whitespace

  <!-- comment -->
  <!---->
`;
const sourceBadSyntax = '<div id="" class=">test</span>';
let source = '';

beforeAll(async () => {
  source = await readFile(componentPath, 'utf8');
});

describe('Svelte markup preprocessor', () => {
  it('processes a simple component', async () => {
    expect.assertions(4);
    const output = preprocess(sourceSimple, preprocessOpts);
    await expect(output).resolves.toBeDefined();
    const result = (await output).toString();
    expect(result).toMatch('> <');
    expect(result).not.toMatch('><');
    expect(result).toMatchSnapshot();
  });

  it('processes a simple component unsafe', async () => {
    expect.assertions(4);
    const output = preprocess(sourceSimple, preprocessOptsUnsafe);
    await expect(output).resolves.toBeDefined();
    const result = (await output).toString();
    expect(result).toMatch('><');
    expect(result).not.toMatch('> <');
    expect(result).toMatchSnapshot();
  });

  it('processes a component', async () => {
    expect.assertions(2);
    const output = preprocess(source, preprocessOpts);
    await expect(output).resolves.toBeDefined();
    const result = (await output).toString();
    expect(result).toMatchSnapshot();
  });

  it('creates and mounts a component', async () => {
    expect.assertions(1);
    const processed = await preprocess(source, preprocessOpts);
    const TestComponent = create(processed.toString(), svelteOpts);
    const target = document.createElement('div');
    new TestComponent({ target });
    expect(target.innerHTML).toMatchSnapshot();
  });

  it('unsafe option makes output smaller', async () => {
    expect.assertions(1);
    const [unsafe, safe] = await Promise.all([
      preprocess(source, preprocessOptsUnsafe),
      preprocess(source, preprocessOpts),
    ]);
    expect(unsafe.toString().length).toBeLessThan(safe.toString().length);
  });

  it('prints error on bad HTML syntax', async () => {
    expect.assertions(1);
    const spy = jest.spyOn(process.stderr, 'write').mockImplementation(() => {});
    await preprocess(sourceBadSyntax, preprocessOpts);
    expect(spy).toHaveBeenCalled();
    spy.mockReset();
    spy.mockRestore();
  });
});
