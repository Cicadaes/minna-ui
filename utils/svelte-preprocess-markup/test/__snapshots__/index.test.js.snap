// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Svelte markup preprocessor creates and mounts a component 1`] = `"<div class=\\"wrapper\\"><a href=\\"\\"><img src=\\"#\\"></a>  <div class=\\"text\\">test Elon Musk</div> <div class=\\"text\\">test <span>ksuM nolE</span></div></div>"`;

exports[`Svelte markup preprocessor processes a component 1`] = `
"<!--
  TEST COMPONENT

  A simple Svelte component for use in tests.

  NOTE:
    In the code, object properties starting with __ are safe to mangle; the name
    can be shortened at build time for smaller JavaScript bundles.

--> <div class=\\"wrapper\\"> <a href=\\"\\"> <img src=\\"#\\"> </a> <!-- Yo yo comment --> <div class=\\"text\\" ref:__target> test {__name} </div> <div class=\\"text\\" ref:__nameReversed> test <span>{__reversed}</span> </div> </div> <script> export default {
    data: () => ({
      __name: 'Elon Musk',
    }),
    computed: {
      __reversed: ({ __name }) => __name.split('').reverse().join(''),
    },
  }; </script> <style>/*discarded*/</style> "
`;

exports[`Svelte markup preprocessor processes a simple component 1`] = `" <div id=\\"target\\"> test </div> <hr class=\\"one two three\\"> <input type=\\"text\\" name=\\"test\\" class=\\"\\" readonly=\\"readonly\\"> whitespace <!-- comment --> <!----> "`;

exports[`Svelte markup preprocessor processes a simple component unsafe 1`] = `"<div id=\\"target\\">test</div><hr class=\\"one three two\\"><input class=\\"\\" name=\\"test\\" readonly> whitespace<!-- comment --><!---->"`;
